File: template/PxCommand/python/template.py
import maya.api.OpenMaya as om

import maya.cmds as cmds


def maya_useNewAPI():
    """
    The presence of this function tells Maya that the plugin produces, and
    expects to be passed, objects created using the Maya Python API 2.0.
    """
    pass



class {{plugin_name}}Command(om.M{{detail_name}}):

    COMMAND_NAME = "{{plugin_name}}Command"
    
    VERSION_FLAG = ["-v", "-version"]
    
    """
    translation command example
    """
    TRANSLATE_FLAG = ["-t", "-translation", (om.MSyntax.kDouble,om.MSyntax.kDouble,om.MSyntax.kDouble)]
    
    NAME_FLAG = ["-n","-name"]
    """
    translation command example
    """
    def __init__(self):
        super({{plugin_name}}Command, self).__init__()
        self.undoable = False
        
    def doIt(self, arg_list):
        
        try:
            arg_db = om.MArgDatabase(self.syntax(), arg_list)
        except:
            self.displayError("Error parsing arguments")
            raise
        
        """
        translation command example
        """
        selection_list = arg_db.getObjectList()

        self.selected_obj = selection_list.getDependNode(0)
        # check if the selection node is transform node or not
        if self.selected_obj.apiType() != om.MFn.kTransform:
            raise RuntimeError("This command requires a transform node")
        
        self.edit = arg_db.isEdit
        self.query = arg_db.isQuery
        
        self.translate = arg_db.isFlagSet({{plugin_name}}Command.TRANSLATE_FLAG[0])
        if self.translate:
            transform_fn = om.MFnTransform(self.selected_obj)
            #get translate in the current space, this is a translate state (undo)
            self.orig_translation = transform_fn.translation(om.MSpace.kTransform)
            
            if self.edit:
                self.new_translation = [arg_db.flagArgumentDouble({{plugin_name}}Command.TRANSLATE_FLAG[0],0),
                                        arg_db.flagArgumentDouble({{plugin_name}}Command.TRANSLATE_FLAG[0],1),
                                        arg_db.flagArgumentDouble({{plugin_name}}Command.TRANSLATE_FLAG[0],2)]
                
                self.undoable = True
            
        self.version = arg_db.isFlagSet({{plugin_name}}Command.VERSION_FLAG[0])
        """
        translation command example
        """
        self.redoIt()
            
    def undoIt(self):
        """
        translation command example
        """
        # will record each stack obj state
        transform_fn =  om.MFnTransform(self.selected_obj)
        # so it will always relate to that stack attr state value
        transform_fn.setTranslation(om.MVector(self.orig_translation),om.MSpace.kTransform)
        """
        translation command example
        """

    def redoIt(self):
        """
        translation command example
        """
        transform_fn = om.MFnTransform(self.selected_obj)
        if self.query:
            if self.translate:
                self.setResult(self.orig_translation)
            else:
                raise RuntimeError("Flag does not support query")
                
        elif self.edit:
            if self.translate:
                # convert to a vector type (c++ binding)
                transform_fn.setTranslation(om.MVector(self.new_translation),om.MSpace.kTransform)
            else:
                raise RuntimeError("Flag does not support edit")
        elif self.version:
            self.setResult("1.0.0")
        else:
            self.setResult(transform_fn.name())
        """
        translation command example
        """
    def isUndoable(self):
        return self.undoable

    @classmethod
    def creator(cls):
        return {{plugin_name}}Command()

    @classmethod
    def create_syntax(cls):

        syntax = om.MSyntax()
        syntax.enableEdit = True
        syntax.enableQuery = True
        """
        translation command example
        """
        # Add flags here
        # pointing to the args list ,optimized way
        syntax.addFlag(*cls.VERSION_FLAG)
        syntax.addFlag(*cls.TRANSLATE_FLAG)
        """
        translation command example
        """
        
        """
        translation command example
        """
        syntax.setObjectType(om.MSyntax.kSelectionList,1,1)
        syntax.useSelectionAsDefault(True)
        """
        translation command example
        """
     
        return syntax



def initializePlugin(plugin):
    """
    """
    vendor = "{{author}}"
    version = "{{version}}"

    plugin_fn = om.MFnPlugin(plugin, vendor, version)
    try:
        plugin_fn.registerCommand({{plugin_name}}Command.COMMAND_NAME, {{plugin_name}}Command.creator, {{plugin_name}}Command.create_syntax)
    except:
        om.MGlobal.displayError("Failed to register command: {0}".format({{plugin_name}}Command.COMMAND_NAME))


def uninitializePlugin(plugin):
    """
    """
    plugin_fn = om.MFnPlugin(plugin)
    try:
        plugin_fn.deregisterCommand({{plugin_name}}Command.COMMAND_NAME)
    except:
        om.MGlobal.displayError("Failed to deregister command: {0}".format({{plugin_name}}Command.COMMAND_NAME))


if __name__ == "__main__":

    cmds.file(new=True, force=True)
    """
    translation command example
    """
    plugin_name = "{{plugin_name}}Command.py"
    cmds.evalDeferred('if cmds.pluginInfo("{0}", q=True, loaded=True): cmds.unloadPlugin("{0}")'.format(plugin_name))
    cmds.evalDeferred('if not cmds.pluginInfo("{0}", q=True, loaded=True): cmds.loadPlugin("{0}")'.format(plugin_name))
    
    cmds.evalDeferred('cmds.polyCube()')
    """
    translation command example
    """

File: template/PxCommand/cpp/template.cpp
#include "{{plugin_name}}Command.h"


//-----------------------------------------------------------------------------
// STATIC CONSTANTS
//-----------------------------------------------------------------------------

static const MString MEL_COMMAND = "{{plugin_name}}Command";
//-----------------------------------------------------------------------------
// translation command example
//-----------------------------------------------------------------------------
static const char* TRANSLATE_FLAG[2] = {"-t","-translate"};
static const char* VERSION_FLAG[2] = {"-v","-version"};
//-----------------------------------------------------------------------------
// translation command example
//-----------------------------------------------------------------------------


//-----------------------------------------------------------------------------
// PUBLIC METHODS
//-----------------------------------------------------------------------------
{{plugin_name}}Command::{{plugin_name}}Command() :
    M{{detail_name}}(),
    mUndoable(false),
    mEdit(false),
    mQuery(false),
    mTranslate(false),
    mVersion(false)
{


}

{{plugin_name}}Command::~{{plugin_name}}Command()
{
}

MStatus {{plugin_name}}Command::doIt(const MArgList& args)
{
    MStatus status;

    MArgDatabase argData(syntax(), args, &status);

    
    if (!status)
    {
		MGlobal::displayError("Failed to parse arguments: " + status.errorString());
        return(status);
	}
    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
    MSelectionList selectionList;
    status = argData.getObjects(selectionList);

    if (!status)
	{ 
        MGlobal::displayError("Failed to get selection list: " + status.errorString());
		return(status);
    }

    // pass the first object in the selection list to the mSelectionObj
    selectionList.getDependNode(0, mSelectionObj);

    if (mSelectionObj.apiType() != MFn::kTransform)
    {
		MGlobal::displayError("Selected object is not a transform node");
		return(MStatus::kFailure);
	}

    mEdit = argData.isEdit();
    mQuery = argData.isQuery();

    mVersion = argData.isFlagSet(VERSION_FLAG[0]);
    mTranslate = argData.isFlagSet(TRANSLATE_FLAG[0]);
    
    if (mTranslate)
    {
        MFnTransform transformFn(mSelectionObj);

        mOrigTranslation = transformFn.getTranslation(MSpace::kTransform);

        if (mEdit)
        {   
            // set the new translation,get the values from the flag based on the index
            mNewTranslation = MVector(argData.flagArgumentDouble(TRANSLATE_FLAG[0], 0),
                									  argData.flagArgumentDouble(TRANSLATE_FLAG[0], 1),
                									  argData.flagArgumentDouble(TRANSLATE_FLAG[0], 2));

            mUndoable = true;

        }
    }
    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------

    return(redoIt());
}

MStatus {{plugin_name}}Command::undoIt()
{   
    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
   MFnTransform transformFn(mSelectionObj);


   transformFn.setTranslation(mOrigTranslation, MSpace::kTransform);

    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
   return MS::kSuccess;
}

MStatus {{plugin_name}}Command::redoIt()
{   
    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
    MFnTransform transformFn(mSelectionObj);

    if (mQuery)
    {
        if (mTranslate)
        {
            double result[3];
            // store the original translation values in the result array
            mOrigTranslation.get(result);


            setResult(MDoubleArray(result,3));

        }
        else
        {
            MGlobal::displayError("Flag does not support query");
            return MS::kFailure;
        }
    }
    else if (mEdit)
    {   
        // set the new translation values if the translate flag is set
        if (mTranslate)
        {
            transformFn.setTranslation(mNewTranslation, MSpace::kTransform);
        }
        else
        {
            MGlobal::displayError("Flag does not support edit");
            return MS::kFailure;
        }
    }
    else if (mVersion)
    {
		setResult("1.0.0");
	}
    else
    {
        setResult(transformFn.name());
	}

    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
    return MS::kSuccess;
}

bool {{plugin_name}}Command::isUndoable() const
{
    return mUndoable;
}


//-----------------------------------------------------------------------------
// STATIC METHODS
//-----------------------------------------------------------------------------
void* {{plugin_name}}Command::Creator()
{
    return(new {{plugin_name}}Command());
}

MString {{plugin_name}}Command::CommandName()
{
    return(MEL_COMMAND);
}

MSyntax {{plugin_name}}Command::CreateSyntax()
{
    MSyntax syntax;

    syntax.enableEdit(true);
    syntax.enableQuery(true);
    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
    syntax.addFlag(TRANSLATE_FLAG[0], TRANSLATE_FLAG[1], MSyntax::kDouble, MSyntax::kDouble, MSyntax::kDouble);
    syntax.addFlag(VERSION_FLAG[0], VERSION_FLAG[1]);

    // can only accept one selected object
    syntax.setObjectType(MSyntax::kSelectionList, 1, 1);
    // will use the current selection list
    syntax.useSelectionAsDefault(true);
    //-----------------------------------------------------------------------------
    // translation command example
    //-----------------------------------------------------------------------------
    return syntax;
}
